File: ride-sharing-app/data/rider.ts
import Ride, { RIDESTATUS } from "./ride";


export default class Rider {

  public name: string;
  public rides: Ride[];
  public id: number;
  public withdrawnRides: number; // to store withdrawn ride count
  public isRiderBusy: boolean;
  constructor(name: string){
    this.name = name;
    this.id = Math.round(Math.random()*10);
    this.rides = [];
    this.isRiderBusy = false;
  }

  public createRide(origin: number, destination: number, noofRiders: number): number {
    if(this.isRiderBusy){
      throw new Error('Cannot book two rides simultaneously');
    }
    let ride = new Ride(Math.round(Math.random()*10), origin, destination, noofRiders);
    this.rides.push(ride);
    this.isRiderBusy = true;
    return ride.id;
  }

  public updateRide(rideUpdate: Ride): void {
    let ride = this.getRide(rideUpdate.id);
    if(!ride){
      throw new Error("Invalid ride");
    }

    //update ride 
    ride.id = rideUpdate.id;
    ride.origin = rideUpdate.origin;
    ride.destination = rideUpdate.destination;
    ride.noOfRiders = rideUpdate.noOfRiders;
  }
  public closeRide(id: number): number {
    //check for invalid id or invalid ride todo 
    let ride = this.getRide(id);
    if(!ride || ride.rideStatus === RIDESTATUS.COMPLETED || ride.rideStatus === RIDESTATUS.WITHDRAWN){
      throw new Error('Ride does not exist or has been completed or withdrawm')
    }
    this.isRiderBusy = false;
    return ride.closeRide(this.rides.length - this.withdrawnRides >10);
  }

  public withdrawRide(id: number): void {
    let ride = this.getRide(id);
    if(!ride || ride.rideStatus === RIDESTATUS.COMPLETED || ride.rideStatus === RIDESTATUS.WITHDRAWN){
      throw new Error(`Ride is invalid or ${ride.rideStatus.toLowerCase()}`)
    }
    this.withdrawnRides = this.withdrawnRides +1;
    ride.rideStatus = RIDESTATUS.WITHDRAWN;
  }

  private getRide(rideId: number): Ride | undefined {
    let ride = this.rides.filter((ride) =>{
      return ride.id === rideId;
    });
    return ride.length > 0 ? ride[0] : undefined
  }
}

File: ride-sharing-app/go/rider.go
package main

import (
	"errors"
	"fmt"
)
type Rider struct{
  Id int
  Name string
  withdrawnRides int
  Rides []Ride
  IsRiderAvailable bool
}

func (rider *Rider) CreateRide(source int, destination int, noOfRiders int, id int) (Ride, error){
  if !rider.IsRiderAvailable {
    return  Ride{}, errors.New("Cannot create multiple rides at a time")
  }
  ride := Ride {
    Id: id,
    Source: source,
    Destination: destination,
    noOfRiders: noOfRiders,
    RideStatus: SCHEDULED,
  }
  rider.Rides = append(rider.Rides, ride)
  return ride, nil
}

func (rider *Rider) UpdateRide(ride Ride) (bool, error){
  rideReturned, err := rider.getRide(ride.Id)
  if(err != nil){
    return false, err
  }
  rideReturned.Source = ride.Source
  rideReturned.Destination = ride.Destination
  rideReturned.noOfRiders = ride.noOfRiders
  return true, nil
}


func (rider *Rider) WithDrawnRide(id int) (bool, error) {
  returnedRide, err := rider.getRide(id)
  if(err != nil){
    return false, err
  }
  if returnedRide.RideStatus != SCHEDULED {
    return false, errors.New("Cannot withdrawn ride, not scheduled")
  }
  rider.withdrawnRides = rider.withdrawnRides + 1
  returnedRide.RideStatus = WITHDRAWN
  rider.IsRiderAvailable = true
  return true, nil;
}

func (rider *Rider) CloseRide(id int) (float32, error) {
  returnedRide, err := rider.getRide(id)
  if(err != nil){
    return 0.0, err
  }
  if returnedRide.RideStatus != SCHEDULED {
    return 0.0, errors.New("Invalid ride to close")
  }
  returnedRide.RideStatus = COMPLETED
  rider.IsRiderAvailable = true
  return  returnedRide.CloseRide(len(rider.Rides) - rider.withdrawnRides >10), nil
}

func (rider *Rider) getRide(id int) (Ride, error) {
  for i:=0; i< len(rider.Rides); i++{
    if rider.Rides[i].Id == id{
      return rider.Rides[i], nil
    }
  }
  return Ride{}, errors.New("Ride not found")
}
